plugins/uber-jar/src/uberjar/JarUtils.scala

```
@@ -0,0 +1,105 @@
+    Files.walkFileTree(root, new SimpleFileVisitor[Path]() {
+      override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        // when we put part of compiled classes, zip already contains some entries. We should not duplicate them.
+        try {
+          out.putNextEntry(new ZipEntry(root.relativize(file).toString))
+          Files.copy(file, out)
+          out.closeEntry()
+        } catch {
+          case e: ZipException => log(s"Failed to add entry, skipping cause: $e")
+        }
+        FileVisitResult.CONTINUE
+      }
+
+      override def preVisitDirectory(dir: Path, attrs: BasicFileAttributes): FileVisitResult = {
+        // when we put part compiled classes, zip already contains some entries. We should not duplicate them.
+        try {
+          out.putNextEntry(new ZipEntry(root.relativize(dir).toString + pathSeparator))
+          out.closeEntry()
+        } catch {
+          case e: ZipException => log(s"Failed to add entry, skipping cause: $e")
+        }
+        FileVisitResult.CONTINUE
+      }
+    })
+  }
+
+  /**
+    * Extracts jars, and writes them on disk. Returns root directory of extracted jars
+    * TODO: in future we probably should save extracted jars in target directory, to reuse them on second run
+    * @param jars list of *.jar files
+    * @param log logger
+    * @return root directory of extracted jars
+    */
+  protected def extractJars(jars: Seq[File])(log: String => Unit): Path = {
+    val destDir = {
+      val path = Files.createTempDirectory("unjars")
+      path.toFile.deleteOnExit()
+      log(s"Unjars directory: $path")
+      path
+    }
+    jars foreach { jar => extractJar(jar, destDir)(log) }
+    destDir
+  }
+
+  /**
+    * Extracts content of single jar file to destination directory.
+    * When extracting jar, if same file already exists, we skip(don't write) this file.
+    * TODO: maybe skipping duplicates is not best strategy. Figure out duplicate strategy.
```

why don't we handle conflicts like this: as part of the uberJar process we let the user provide function from a list of pairs (jar file, contained file) and they have to give us back a list of pairs should be included. We throw an exception if it contains duplicates. Default implementation either `identity` (meaning we fail) or something like `_.filterNot(_._2 == "Manifest....")`. Better to fail and require solution than silently discard things during conflicts.